src/lib/x509/x509_ext.cpp

```diff
@@ -847,6 +856,518 @@ void TNAuthList::decode_inner(const std::vector<uint8_t>& in) {
+         rdi_obj_ber.decode(rdi);
+         m_rdi = rdi;
+      } else if(rdi_type_tag != ASN1_Type::NoObject) {
+         throw Decoding_Error(fmt("Unexpected type for ASIdentifiers rdi: {}", static_cast<uint32_t>(rdi_type_tag)));
+      }
+   }
+
+   // just an rdi
+   if(elem_type_tag == 1) {
+      BER_Decoder rdi_obj_ber = BER_Decoder(elem_obj);
+      ASIdentifierChoice rdi;
+      rdi_obj_ber.decode(rdi);
+      m_rdi = rdi;
+      BER_Object end = obj_ber.get_next_object();
+      ASN1_Type end_type_tag = end.type_tag();
+      if(end_type_tag != ASN1_Type::NoObject) {
+         throw Decoding_Error(
+            fmt("Unexpected element with type {} in ASIdentifiers", static_cast<uint32_t>(end_type_tag)));
+      }
+   }
+}
+
+void ASBlocks::ASIdentifierChoice::encode_into(Botan::DER_Encoder& into) const {
+   if(this->m_as_ranges.has_value()) {
+      into.start_sequence().encode_list(this->m_as_ranges.value()).end_cons();
+   } else {
+      into.encode_null();
+   }
+}
+
+void ASBlocks::ASIdentifierChoice::decode_from(Botan::BER_Decoder& from) {
+   BER_Object obj = from.get_next_object();
+   ASN1_Type type_tag = obj.type_tag();
+
+   if(type_tag == ASN1_Type::Null) {
+      m_as_ranges = std::nullopt;
+   } else if(type_tag == ASN1_Type::Sequence) {
+      BER_Decoder obj_ber = BER_Decoder(obj);
+      std::vector<ASIdOrRange> as_ranges;
+
+      while(obj_ber.more_items()) {
+         ASIdOrRange as_id_or_range;
+         obj_ber.decode(as_id_or_range);
+         as_ranges.push_back(as_id_or_range);
+      }
+
+      m_as_ranges = as_ranges;
+   } else {
+      throw Decoding_Error(fmt("Unexpected type for ASIdentifierChoice {}", static_cast<uint32_t>(type_tag)));
+   }
+}
+
+std::vector<uint8_t> ASBlocks::ASIdOrRange::encode_asnum(asnum_t asnum) const {
+   std::vector<uint8_t> bytes;
+   for(size_t i = 0; i < sizeof(asnum); i++) {
+      bytes.push_back(static_cast<uint8_t>((asnum >> 8 * (sizeof(asnum) - i - 1)) & 0xff));
+   }
+
+   // remove leading zeros
+   while(bytes[0] == 0 && bytes.size() > 1) {
+      bytes.erase(bytes.begin());
+   }
+
+   // add one padding of zero to not get it confused as a negative number
+   if(((bytes[0] >> 7) & 1) == 1) {
+      bytes.insert(bytes.begin(), 0);
+   }
+
+   return bytes;
+}
+
+void ASBlocks::ASIdOrRange::encode_into(Botan::DER_Encoder& into) const {
+   if(m_min == m_max) {
+      into.add_object(ASN1_Type::Integer, ASN1_Class::Universal, encode_asnum(m_min));
+   } else {
+      if(m_min >= m_max) {
+         throw Encoding_Error("AS range numbers must be sorted");
+      }
+      into.start_sequence()
+         .add_object(ASN1_Type::Integer, ASN1_Class::Universal, encode_asnum(m_min))
+         .add_object(ASN1_Type::Integer, ASN1_Class::Universal, encode_asnum(m_max))
+         .end_cons();
+   }
+}
+
+void ASBlocks::ASIdOrRange::decode_from(BER_Decoder& from) {
+   BER_Object obj = from.get_next_object();
+   ASN1_Type type_tag = obj.type_tag();
```

Comment:
I think this function can be quite a bit simpler overall, something like (untested)

```
ASN1_Tag next_tag = from.peek_next_object().type_tag();
if(next_tag == ASN1_Type::Integer) {
   from.decode(m_min);
   m_max = m_min;
} else if(next_tag == ASN11_Type::Sequence) {
  from.start_sequence().decode(m_min).decode(m_max).end_cons();
} else {
   throw ...
}
```